spec/lib/state_machine_spec.rb

```diff
@@ -0,0 +1,210 @@
+    expect(test.can_make_second_state?).to be false
+    expect(test.can_make_third_state?).to be true
+    
+    expect(test.make_third_state).to be true
+
+    expect(test.is_first_state?).to be false
+    expect(test.is_second_state?).to be false
+    expect(test.is_third_state?).to be true
+
+    expect(test.can_make_first_state?).to be false
+    expect(test.can_make_second_state?).to be false
+    expect(test.can_make_third_state?).to be false
+  end
+
+  it "test second_state_at gets filled in" do
+    test = StateMachineTestClass.create
+    expect(test.has_been_made_second_state?).to be false
+    expect(test.second_state_at).to be_nil
+    expect(test.make_second_state).to be true
+    expect(test.second_state_at).not_to be_nil
+
+    expect(test.has_been_made_second_state?).to be true
+    expect{ test.has_been_made_third_state? }.to raise_error(NotImplementedError)
+  end
+
+  it "test state machine callbacks" do
+    test = StateMachineTestClass.create
+
+    expect(test.make_second_state).to be true
+    expect(test.callbacks_happened.count).to eq(1)
+    expect(test.callbacks_happened[:second_state].count).to eq(2)
+    expect(test.callbacks_happened[:second_state][:before][:first_state]).to eq(2)
+    expect(test.callbacks_happened[:second_state][:after][:first_state]).to eq(1)
+  end
+
+  it "test state machine callbacks don't fire if state didn't change" do
+    test = StateMachineTestClass.create
+    expect(test.make_second_state).to be true
+    first_time = test.callbacks_happened
+
+    expect(first_time.present?).to be true
+
+    expect(test.make_second_state).to be true
+    test.save
+    expect(test.callbacks_happened).to eq(first_time)
+  end
+
+  it "test state machine validation errors" do
+    test = StateMachineTestClass.create
+
+    test.state = "bad_state"
+    expect(test.save).to be false
+    expect(test.errors[:state].first).to eq("bad_state is not a valid state.")
+
+    test = StateMachineTestClass.create
+    expect(test.make_second_state).to be true
+    expect(test.make_first_state).to be false
+    expect(test.errors[:state].first).to eq("Cannot transition from second_state to first_state.")
+
+    test = StateMachineTestClass.create
+    expect(test.make_third_state).to be true
+    expect(test.make_fourth_state).to be false
+  end
+
+  it "test state machine validation errors as exceptions" do
+    test = StateMachineTestClass.create
+    expect(test.make_second_state).to be true
+    expect{test.make_first_state!}.to raise_error(Exception)
+
+    test = StateMachineTestClass.create
+    expect(test.make_third_state).to be true
+
+    expect{test.make_fourth_state!}.to raise_error(Exception)
+  end
+
+  it "test overwriting timestamps" do
+    test = StateMachineTestClass.create
+    expect(test.second_state_at).to be_nil
+    expect(test.make_second_state).to be true
+    expect(test.second_state_at).not_to eq(nil)
+
+    was = test.second_state_at
+    Timecop.travel(Time.now + 2)
```

Comment:
should add comment for the amount of time 2 is actually adding (seconds/milliseconds)